File: WebPrototype/js/levelManager.js
// Same responsibilities as Assets/Scripts/Level/LevelManager.cs: level
// lookup and unlock rules. Takes progress as a parameter rather than owning
// it, same reasoning as the Unity version - persistence is save.js's job.
function getLevel(levelId) {
  return LEVELS.find(level => level.levelId === levelId) || null;
}

function getNextLevel(currentLevelId) {
  return getLevel(currentLevelId + 1);
}

function isLevelUnlocked(progress, levelId) {
  return progress.unlockedLevels.includes(levelId);
}

function unlockLevel(progress, levelId) {
  if (!progress.unlockedLevels.includes(levelId)) {
    progress.unlockedLevels.push(levelId);
  }
}

function unlockNextLevel(progress, completedLevelId) {
  const nextLevel = getNextLevel(completedLevelId);

  if (nextLevel) {
    unlockLevel(progress, nextLevel.levelId);
  }
}

File: CSharpEngineerQuest/Assets/Scripts/Level/LevelManager.cs
using System.Collections.Generic;
using System.Linq;
using CSharpEngineerQuest.Data;

namespace CSharpEngineerQuest.Level
{
    /// <summary>
    /// Looks up level content and decides unlock rules. Does not own or persist
    /// PlayerProgress - that belongs to SaveManager/PlayerProgressManager (later
    /// phases), which call in here and pass their PlayerProgress instance.
    /// </summary>
    public class LevelManager
    {
        private readonly List<LevelData> _levels;

        public LevelData CurrentLevel { get; private set; }

        public LevelManager(LevelLoader levelLoader)
        {
            _levels = levelLoader.LoadAllLevels();
        }

        public LevelData GetLevel(int levelId)
        {
            return _levels.FirstOrDefault(level => level.levelId == levelId);
        }

        /// <summary>
        /// Levels are sequential by levelId for the MVP's linear world map.
        /// </summary>
        public LevelData GetNextLevel(int currentLevelId)
        {
            return GetLevel(currentLevelId + 1);
        }

        public void SetCurrentLevel(int levelId)
        {
            CurrentLevel = GetLevel(levelId);
        }

        public bool IsLevelUnlocked(PlayerProgress progress, int levelId)
        {
            return progress.unlockedLevels.Contains(levelId);
        }

        public void UnlockLevel(PlayerProgress progress, int levelId)
        {
            if (!progress.unlockedLevels.Contains(levelId))
            {
                progress.unlockedLevels.Add(levelId);
            }
        }

        /// <summary>
        /// Call after a level is completed to unlock whatever comes after it.
        /// Does nothing if there is no next level yet (e.g. last authored level).
        /// </summary>
        public void UnlockNextLevel(PlayerProgress progress, int completedLevelId)
        {
            LevelData nextLevel = GetNextLevel(completedLevelId);
            if (nextLevel != null)
            {
                UnlockLevel(progress, nextLevel.levelId);
            }
        }
    }
}
